tools/timeoutcmd/timeoutcmd.go

```diff
@@ -46,73 +43,51 @@ func (c *Command) AppendEnv(env string) {
 
 // SetStandardIO sets the input and outputs of the command.
 func (c *Command) SetStandardIO(in io.Reader, out, err io.Writer) {
-	c.cmd.Stdin = in
-	c.cmd.Stdout = out
-	c.cmd.Stderr = err
+	c.cmd.Stdin, c.cmd.Stdout, c.cmd.Stderr = in, out, err
 }
 
 // Start starts the command run.
 func (c *Command) Start() error {
-	if c.timeout > 0 {
-		// Setpgid: true creates a new process group for cmd and its subprocesses
-		// this way we can kill the whole process group
-		c.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
-	}
-
+	// setting up notification for signals so we can have
+	// separated logic to end the process
+	interruptChan := make(chan os.Signal)
+	signal.Notify(interruptChan, os.Interrupt, os.Kill)
+	var interrupted bool
+	go func() {
+		<-interruptChan
+		interrupted = true
+	}()
+
+	// start the process
 	if err := c.cmd.Start(); err != nil {
 		return err
 	}
 
-	if c.timeout > 0 {
-		// terminate the process after the given timeout
-		c.timeoutTimer = time.AfterFunc(c.timeout, func() {
-			if err := c.Stop(); err != nil {
-				log.Warnf("Failed to kill the process, error: %s", err)
-			}
-		})
-
-		// Setpgid: true creates a new process group for cmd and its subprocesses
-		// this way cmd will not belong to its parent process group,
-		// cmd will not be killed when you hit ^C in your terminal
-		// to fix this, we listen and handle Interrupt signal manually
-		c.interruptChan = make(chan os.Signal, 1)
-		signal.Notify(c.interruptChan, os.Interrupt)
-		go func() {
-			<-c.interruptChan
-			signal.Stop(c.interruptChan)
-			if err := c.Stop(); err != nil {
-				log.Warnf("Failed to kill the process, error: %s", err)
-			}
-		}()
-	}
-
-	return c.cmd.Wait()
-}
-
-// Stop terminates the command run.
-func (c *Command) Stop() error {
-	if c.cmd.Process == nil {
-		// not yet started
-		return nil
-	}
+	// Wait for the process to finish
+	done := make(chan error, 1)
+	go func() {
+		done <- c.cmd.Wait()
+	}()
 
-	pid := c.cmd.Process.Pid
+	// or kill it after a timeout (whichever happens first)
+	var timeoutChan <-chan time.Time
 	if c.timeout > 0 {
-		// stop listening on os.Interrupt signal
-		signal.Stop(c.interruptChan)
-		// stop the timeout timer
-		c.timeoutTimer.Stop()
+		timeoutChan = time.After(c.timeout)
+	}
 
-		// use the negative process group id, to kill the whole process group
-		pgid, err := syscall.Getpgid(c.cmd.Process.Pid)
-		if err != nil {
-			return err
+	// exiting the method for the two supported cases: finish/error or timeout
+	select {
```

Comment:
why do you introduce `interrupted` ? why do not you just listen for the signal channel as well?

```
	interruptChan := make(chan os.Signal)
	signal.Notify(interruptChan, os.Interrupt, os.Kill)

	var timeoutChan <-chan time.Time
	timeoutChan = time.After(5 * time.Second)

	done := make(chan error, 1)

	select {
	case <-interruptChan:
		fmt.Println("interrupted")
	case <-timeoutChan:
		fmt.Println("timeout")
	case <-done:
		fmt.Println("finished")
	}
```

Reply:
Because this select block the method before it is finished. If I wire interruptChan here then my method will exit instantly, this way I won't be able to handle the error message coming from the command that we are waiting for.